src/developer_opts.c

```diff
@@ -1508,7 +1560,60 @@ void dev_output_pgm_defs(char *pgmidcp) {
         );
       }
 
+    if(udev && pgm->usbpid && (pgm->conntype == CONNTYPE_USB || is_serialadapter(pgm))) {
+      void (* pi)(PROGRAMMER *) = pgm->initpgm;
+      const char *ids = cache_string(str_ccpgmids(pgm->id));
+      int usbvid = pgm->usbvid, ishid =
+        pi == jtag3_initpgm || pi == jtag3_pdi_initpgm || pi == jtag3_updi_initpgm ||
+        pi == jtag3_dw_initpgm || pi == stk500v2_jtag3_initpgm || pi == jtag3_tpi_initpgm;
+
+      if(!lfirst(pgm->usbpid)) {
+        if(pi == flip1_initpgm || pi == flip2_initpgm) { // Bootloaders, add possible part pids
+          for(LNODEID lp = lfirst(part_list); lp; lp = lnext(lp)) {
+            AVRPART *pt = ldata(lp);
+            if(pt->usbpid)
+              udr = add_udev(udr, &ui, usbvid, pt->usbpid, 0, ids);
+          }
+        }
+      }
+
+      for(LNODEID pidn=lfirst(pgm->usbpid); pidn; pidn=lnext(pidn))
+        udr = add_udev(udr, &ui, usbvid, *(int *) ldata(pidn), ishid, ids);
+    }
+
     if(raw)
       dev_pgm_raw(pgm);
   }
+
+  if(udev && ui) {
+    dev_info("# 1. Put Linux udev rules into /etc/udev/rules.d/80-avrdude.rules (or similar)\n");
+    dev_info("# 2. sudo udevadm control --reload-rules && sudo udevadm trigger (or similar)\n");
+    dev_info("# 3. Unplug the device and plug it in again\n");
```

Comment:
I have written down a few things about using udevadm after changing the set of udev fules. Somewhere IIRC for https://github.com/socratools ...

  * IIRC the reload happens automatically, but should be made explicit for programmatic use to avoid race conditions.

  * IIRC just re-triggering ALL devices can cause some annoyances or even problems due at least to the volume of events.

    The alternative is to unplug/replug the device in question, or tell `udevadm` to only re-trigger the events for a positive list of specific devices using a list usb vid/pid tuples.

    That list is easy to make for the three devices socratools supports, but not for avrdude's dozens and dozens and dozens of programmers.

When I have found that information, a few hints could be added to the comments about using `udevadm trigger`. Oh, and you **either** re-trigger the udev events, or you unplug and replug the device. Doing both duplicates effort.